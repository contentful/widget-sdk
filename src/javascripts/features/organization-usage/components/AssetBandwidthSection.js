import React from 'react';
import { css } from 'emotion';
import { partialRight } from 'lodash';
import {
  TextLink,
  Note,
  Typography,
  Heading,
  Paragraph,
} from '@contentful/forma-36-react-components';
import * as tokens from '@contentful/forma-36-tokens';

import { track } from 'analytics/Analytics';
import { shortenStorageUnit } from 'utils/NumberUtils';
import { useUsageState } from '../hooks/usageContext';

const styles = {
  heading: css({
    color: '#536171',
    fontWeight: tokens.fontWeightNormal,
  }),
  usageNumber: css({
    color: '#253545',
    fontSize: tokens.fontSize3Xl,
    fontWeight: tokens.fontWeightMedium,
  }),
  limit: css({
    fontSize: tokens.fontSizeM,
  }),
  includedLimit: css({
    color: '#6A7889',
  }),
  overage: css({
    color: tokens.colorOrangeMid,
  }),
  learnMoreLink: css({
    marginLeft: tokens.spacingS,
  }),
  note: css({
    width: '40%',
    marginTop: tokens.spacingS,
    color: '#27457A',
  }),
};

export const AssetBandwidthSection = () => {
  const { assetBandwidthData } = useUsageState();

  const { limit, usage, uom } = assetBandwidthData;
  const withUnit = partialRight(shortenStorageUnit, uom);
  const handleClick = () => {
    track('usage:fair_use_policy_clicked');
  };
  return (
    <Typography>
      <Heading element="h2" className={styles.heading}>
        Total asset bandwidth
      </Heading>
      <Paragraph data-test-id="asset-bandwidth-usage" className={styles.usageNumber}>
        {withUnit(usage)}
      </Paragraph>
      <div className={styles.limit}>
        <strong data-test-id="asset-bandwidth-limit" className={styles.includedLimit}>{`${withUnit(
          limit
        )} included`}</strong>
        {usage > limit && (
          <small className={styles.overage} data-test-id="asset-bandwidth-overage">
            {' '}
            + {withUnit(usage - limit)} overage
          </small>
        )}
        <TextLink
          href="https://www.contentful.com/r/knowledgebase/fair-use/"
          target="_blank"
          onClick={handleClick}
          data-test-id="fair_use_policy_link"
          rel="noopener noreferrer"
          className={styles.learnMoreLink}>
          Fair Use Policy
        </TextLink>
        <Note className={styles.note}>
          Please note that asset bandwidth is calculated daily with a 48 hour delay.
        </Note>
      </div>
    </Typography>
  );
};
