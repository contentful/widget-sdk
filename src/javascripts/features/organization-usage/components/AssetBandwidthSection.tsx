import React from 'react';
import { css } from 'emotion';
import { partialRight } from 'lodash';
import {
  TextLink,
  Note,
  Typography,
  Heading,
  Paragraph,
  SkeletonContainer,
  SkeletonDisplayText,
  SkeletonBodyText,
} from '@contentful/forma-36-react-components';
import tokens from '@contentful/forma-36-tokens';

import { track } from 'analytics/Analytics';
import { shortenStorageUnit } from 'utils/NumberUtils';
import { useUsageState } from '../hooks/usageContext';

const styles = {
  heading: css({
    color: '#536171',
    fontWeight: tokens.fontWeightNormal,
  }),
  usageNumber: css({
    color: '#253545',
    fontSize: tokens.fontSize3Xl,
    fontWeight: tokens.fontWeightMedium,
  }),
  limit: css({
    fontSize: tokens.fontSizeM,
  }),
  includedLimit: css({
    color: '#6A7889',
  }),
  overage: css({
    color: tokens.colorOrangeMid,
  }),
  learnMoreLink: css({
    marginLeft: tokens.spacingS,
  }),
  note: css({
    width: '40%',
    marginTop: tokens.spacingS,
    color: '#27457A',
  }),
};

export const AssetBandwidthSection = () => {
  const { assetBandwidthData, isLoading } = useUsageState();

  const { limit, usage, uom } = assetBandwidthData ?? {};

  // Show loading state once at the initial data fetch
  const isPageLoading = isLoading && !assetBandwidthData;

  const withUnit = partialRight(shortenStorageUnit, uom);
  const handleClick = () => {
    track('usage:fair_use_policy_clicked', { source: 'Asset Bandwidth' });
  };

  const UsageAndLimit = () => (
    <>
      <Paragraph data-test-id="asset-bandwidth-usage" className={styles.usageNumber}>
        {
          // eslint-disable-next-line @typescript-eslint/ban-ts-comment
          // @ts-ignore
          withUnit(usage)
        }
      </Paragraph>
      <div className={styles.limit}>
        <strong data-test-id="asset-bandwidth-limit" className={styles.includedLimit}>{
          // eslint-disable-next-line @typescript-eslint/ban-ts-comment
          // @ts-ignore
          `${withUnit(limit)} included`
        }</strong>
        {usage > limit && (
          <small className={styles.overage} data-test-id="asset-bandwidth-overage">
            {' '}
            +{' '}
            {
              // eslint-disable-next-line @typescript-eslint/ban-ts-comment
              // @ts-ignore
              withUnit(usage - limit)
            }{' '}
            overage
          </small>
        )}
        <TextLink
          href="https://www.contentful.com/r/knowledgebase/fair-use/"
          target="_blank"
          onClick={handleClick}
          data-test-id="fair_use_policy_link"
          rel="noopener noreferrer"
          className={styles.learnMoreLink}>
          Fair Use Policy
        </TextLink>
      </div>
    </>
  );

  return (
    <Typography>
      <Heading element="h2" className={styles.heading}>
        Total asset bandwidth
      </Heading>
      {isPageLoading && (
        <SkeletonContainer svgWidth="40%" svgHeight="80px">
          <SkeletonDisplayText numberOfLines={1} />
          <SkeletonBodyText numberOfLines={1} offsetTop={50} />
        </SkeletonContainer>
      )}
      {!isPageLoading && <UsageAndLimit />}
      <Note className={styles.note}>
        Please note that asset bandwidth is calculated daily with a 48 hour delay.
      </Note>
    </Typography>
  );
};
