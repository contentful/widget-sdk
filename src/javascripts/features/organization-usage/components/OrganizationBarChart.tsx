import React from 'react';
import { last } from 'lodash';
import tokens from '@contentful/forma-36-tokens';
import { css } from 'emotion';
import { shorten } from 'utils/NumberUtils';
import { useChart } from '../hooks/useChart';
import { useUsageState } from '../hooks/usageContext';

const styles = {
  chartWrapper: css({
    height: '512px',
    width: '100%',
  }),
};

const accumulation = (acc, value) => acc.concat(value + (last(acc) || 0));

export const accumulate = (data) => data.reduce(accumulation, []);

export const applyBorderStyle = (value) => ({
  value,
  itemStyle: { borderWidth: value > 0 ? 2 : 0 },
});

const propsToChartOption = ({ period, usage, includedLimit }) => {
  const dataWithStyle = accumulate(usage).map(applyBorderStyle);

  const markLine = {
    symbol: ['none', 'circle'],
    data: [{ yAxis: includedLimit }],
    lineStyle: {
      color: '#CC3C52',
    },
    label: {
      show: false,
    },
  };

  return {
    xAxis: {
      data: period,
      type: 'category',
      axisLabel: {
        color: '#6A7889',
        fontFamily: tokens.fontStackPrimary,
        fontSize: 14,
      },
      axisLine: {
        lineStyle: {
          color: '#B4C3CA',
        },
      },
    },
    yAxis: {
      scale: false,
      min: 0,
      position: 'right',
      splitLine: {
        lineStyle: {
          color: '#D3DCE0',
          type: 'dashed',
        },
      },
      axisLabel: {
        color: '#536171',
        fontFamily: tokens.fontStackPrimary,
        fontSize: 14,
        formatter: shorten,
      },
      axisLine: {
        lineStyle: {
          color: '#B4C3CA',
        },
      },
    },
    series: [
      {
        data: dataWithStyle,
        name: 'API Requests',
        type: 'bar',
        itemStyle: {
          color: '#2E75D4',
          borderColor: '#2E75D4',
          opacity: 0.5,
        },
        markLine: includedLimit ? markLine : undefined,
      },
    ],
    grid: {
      left: '50px',
      right: '50px',
      bottom: 70,
    },
    dataZoom: [
      {
        id: 'dataZoomX',
        type: 'slider',
      },
      {
        type: 'inside',
        throttle: 50,
        filterMode: 'filter',
      },
    ],
    tooltip: {
      show: true,
      trigger: 'axis',
      backgroundColor: '#192532',
    },
    toolbox: {
      show: true,
      feature: {
        saveAsImage: {
          show: true,
          type: 'png',
          name: `Organization-Usage-${new Date().toDateString()}`,
          title: 'Save as an image',
        },
        magicType: {
          type: ['line', 'bar'],
          title: {
            line: 'Line Chart',
            bar: 'Bar Chart',
          },
        },
      },
    },
  };
};

export const OrganizationBarChart = () => {
  const { periodDates, periodicUsage, apiRequestIncludedLimit, isLoading } = useUsageState();

  const usage = isLoading ? [] : periodicUsage?.org.usage;
  const chartRef = useChart(
    propsToChartOption({
      period: periodDates,
      usage,
      includedLimit: apiRequestIncludedLimit,
    })
  );
  return (
    <div
      ref={chartRef}
      className={styles.chartWrapper}
      data-test-id="organization-usage-bar-chart"
    />
  );
};
