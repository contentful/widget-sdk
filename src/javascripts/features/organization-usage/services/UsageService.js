import { map, sum, unzip } from 'lodash';
import { USAGE_INSIGHTS, getAlphaHeader } from 'alphaHeaders.js';
import * as EndpointFactory from 'data/EndpointFactory';

const headers = getAlphaHeader(USAGE_INSIGHTS);

/**
 * Gets the usage periods
 * @param  {OrganizationEndpoint} endpoint
 * @return {Promise}
 */
export const getPeriods = (endpoint) =>
  endpoint(
    {
      method: 'GET',
      path: ['usage_periods'],
    },
    headers
  );

/**
 * Return organization usage by API type given a date range.
 *
 * @param  {OrganizationEndpoint} endpoint - Usage endpoint
 * @param  {Object} query - Search and/or filter parameters
 * @param  {string} query.startDate - Beginning of search period in yyyy-mm-dd format
 * @param  {string} query.endDate - End of search period in yyyy-mm-dd format
 * @return {Promise}
 */
export const getOrgUsage = (endpoint, { startDate, endDate }) =>
  endpoint({
    method: 'GET',
    path: ['organization_periodic_usages'],
    query: {
      'metric[in]': ['cpa', 'cda', 'cma', 'gql'],
      'dateRange.startAt': startDate,
      'dateRange.endAt': endDate ?? '',
    },
  });

/**
 * Return organization usage by API type given a date range.
 * Available api types are cma, cpa, cda, gql.
 *
 * @param  {OrganizationEndpoint} endpoint - Usage endpoint
 * @param  {Object} query - Search and/or filter parameters
 * @param  {string} query.startDate - Beginning of search period in yyyy-mm-dd format
 * @param  {string} query.endDate - End of search period in yyyy-mm-dd format
 * @param  {string} query.apiType - One of cma, cpa, cda, gql
 * @return {Promise}
 */
export const getApiUsage = (endpoint, { apiType, startDate, endDate, limit }) =>
  endpoint({
    method: 'GET',
    path: ['space_periodic_usages'],
    query: {
      'metric[in]': apiType,
      'dateRange.startAt': startDate,
      'dateRange.endAt': endDate ?? '',
      limit,
    },
  });

export const extractValues = (api) => ({
  ...api,
  usage: Object.values(api.usagePerDay),
});

export const transformApi = (apis) =>
  apis.reduce(
    (acc, { type, api }) => ({
      ...acc,
      [type]: { ...api, items: api.items.map(extractValues) },
    }),
    {}
  );

export const transformOrg = (org) => {
  const newOrg = org.items.reduce(
    (acc, api) => ({
      ...acc,
      [api.metric]: Object.values(api.usagePerDay),
    }),
    {}
  );

  return map(unzip(Object.values(newOrg)), sum);
};

export const mapResponseToState = ({ org, cma, cda, cpa, gql }) => ({
  periodicUsage: {
    org: { usage: transformOrg(org) },
    apis: transformApi([
      { type: 'cma', api: cma },
      { type: 'cda', api: cda },
      { type: 'cpa', api: cpa },
      { type: 'gql', api: gql },
    ]),
  },
});

export const loadPeriodData = async (orgId, period) => {
  const endpoint = EndpointFactory.createOrganizationEndpoint(orgId);
  const promises = [
    getOrgUsage(endpoint, {
      startDate: period.startDate,
      endDate: period.endDate,
    }),
    ...['cma', 'cda', 'cpa', 'gql'].map((apiType) =>
      getApiUsage(endpoint, {
        apiType,
        startDate: period.startDate,
        endDate: period.endDate,
        limit: 5,
      })
    ),
  ];

  const [org, cma, cda, cpa, gql] = await Promise.all(promises);

  return { org, cma, cda, cpa, gql };
};
