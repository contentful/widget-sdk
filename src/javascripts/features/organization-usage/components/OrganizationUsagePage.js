import React from 'react';
import { Tabs, Tab, TabPanel, Heading } from '@contentful/forma-36-react-components';
import { Grid, GridItem } from '@contentful/forma-36-react-components';
import { SpacesTabs } from './SpacesTabs';
import { OrganizationBarChart } from './OrganizationBarChart';
import { OrganizationUsageInfo } from './OrganizationUsageInfo';
import { AssetBandwidthSection } from './AssetBandwidthSection';

import { css } from 'emotion';
import tokens from '@contentful/forma-36-tokens';

import { track } from 'analytics/Analytics';
import { useUsageState, useUsageDispatch } from '../hooks/usageContext';

const styles = {
  tabPanel: css({
    paddingTop: tokens.spacing2Xl,
    paddingBottom: tokens.spacing2Xl,
  }),
  heading: css({
    color: '#6A7889',
    fontWeight: tokens.fontWeightNormal,
    paddingBottom: tokens.spacingXl,
    paddingTop: tokens.spacing2Xl,
  }),
};

export const OrganizationUsagePage = () => {
  const { selectedMainTab, isAssetBandwidthTab } = useUsageState();
  const dispatch = useUsageDispatch();

  const handleSelected = (id) => {
    track('usage:org_tab_selected', { old: selectedMainTab, new: id });
    dispatch({ type: 'SWITCH_MAIN_TAB', value: id });
  };

  return (
    <>
      <Tabs withDivider>
        <Tab id="apiRequest" selected={!isAssetBandwidthTab} onSelect={handleSelected}>
          API Requests
        </Tab>
        <Tab
          testId="organization-usage_asset-bandwidth-tab"
          id="assetBandwidth"
          selected={isAssetBandwidthTab}
          onSelect={handleSelected}>
          Asset Bandwidth
        </Tab>
      </Tabs>

      {!isAssetBandwidthTab && (
        <TabPanel id="apiRequest" className={styles.tabPanel}>
          <Grid columns={'1fr 2fr'}>
            <GridItem>
              <OrganizationUsageInfo />
            </GridItem>
            <GridItem>
              <OrganizationBarChart />
            </GridItem>
          </Grid>
          <Heading element="h2" className={styles.heading}>
            View API requests by type and space
          </Heading>
          <SpacesTabs />
        </TabPanel>
      )}
      {isAssetBandwidthTab && (
        <TabPanel id="assetBandwidth" className={styles.tabPanel}>
          <Grid columns={1}>
            <GridItem>
              <AssetBandwidthSection />
            </GridItem>
          </Grid>
        </TabPanel>
      )}
    </>
  );
};
