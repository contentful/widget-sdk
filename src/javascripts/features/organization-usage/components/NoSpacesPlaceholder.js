import React from 'react';
import EmptyStateContainer, {
  defaultSVGStyle,
} from 'components/EmptyStateContainer/EmptyStateContainer';
import EmptyStateAdminIllustration from 'svg/folder-illustration.svg';
import { Heading, Paragraph, Button, TextLink } from '@contentful/forma-36-react-components';

import * as CreateSpace from 'services/CreateSpace';
import { useUsageState } from '../hooks/usageContext';

export const NoSpacesPlaceholder = () => {
  const { orgId } = useUsageState();
  return (
    <EmptyStateContainer data-test-id="usage-page__no-spaces-placeholder">
      <EmptyStateAdminIllustration className={defaultSVGStyle} />
      <Heading>Looking for your usage data?</Heading>
      <Paragraph>
        Looks like you need to create a space first. Whatâ€™s a space? Good question! A space is an
        area to manage and store content for a specific project.
      </Paragraph>
      <Paragraph>
        You can learn more about spaces in the{' '}
        <TextLink
          href="https://www.contentful.com/developers/docs/concepts/domain-model/"
          target="_blank"
          rel="noopener noreferrer">
          domain model documentation
        </TextLink>
        .
      </Paragraph>
      <Button onClick={() => CreateSpace.showDialog(orgId)}>Add a space</Button>
    </EmptyStateContainer>
  );
};
