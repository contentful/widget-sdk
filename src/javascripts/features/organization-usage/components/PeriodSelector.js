import React from 'react';
import moment from 'moment';
import { Select, Option } from '@contentful/forma-36-react-components';
import { cond, constant, stubTrue } from 'lodash';
import { track } from 'analytics/Analytics';
import { useUsageState, useUsageDispatch } from '../hooks/usageContext';
import { pick } from 'lodash/fp';
import { loadPeriodData } from '../services/UsageService';

const formatDate = (date) => moment(date).format('DD MMM');

export const PeriodSelector = () => {
  const { periods, selectedPeriodIndex, orgId, isTeamOrEnterpriseCustomer } = useUsageState();
  const dispatch = useUsageDispatch();

  const handleChange = async (e) => {
    dispatch({ type: 'SET_LOADING', value: true });

    const newPeriodIndex = parseInt(e.target.value);

    // analytics
    const oldPeriod = periods[selectedPeriodIndex];
    const newPeriod = periods[newPeriodIndex];
    track('usage:period_selected', {
      oldPeriod: pick(['startDate', 'endDate'], oldPeriod),
      newPeriod: pick(['startDate', 'endDate'], newPeriod),
    });

    dispatch({ type: 'CHANGE_PERIOD', value: newPeriodIndex });
    dispatch({ type: 'SWITCH_MAIN_TAB', value: 'apiRequest' });
    const data = await loadPeriodData(orgId, newPeriod);
    dispatch({ type: 'SET_USAGE_DATA', value: data });
    dispatch({ type: 'SET_LOADING', value: false });
  };

  const processDate = (startDate, endDate) => {
    const isCurrentPeriod = endDate === null || moment().diff(moment(endDate), 'days') == 0; // the end day is today
    const start = moment(startDate);
    const end = isCurrentPeriod
      ? moment(start).add(1, 'month').subtract(1, 'day')
      : moment(endDate);
    return `${formatDate(start)} â€“ ${formatDate(end)} ${cond([
      [constant(isCurrentPeriod), constant('(current)')],
      [constant(end.year() === moment().year()), constant(`(${moment().to(start)})`)],
      [constant(stubTrue), constant(end.year())],
    ])()}`;
  };

  const renderSelect = () => {
    return (
      <Select
        defaultValue={selectedPeriodIndex}
        onChange={handleChange}
        width="auto"
        name="period-selector"
        testId="period-selector"
        id="period-selector">
        {periods.map(({ startDate, endDate }, index) => {
          const value = `${index}`;

          return (
            <Option key={index} value={value}>
              {processDate(startDate, endDate)}
            </Option>
          );
        })}
      </Select>
    );
  };

  const renderCurrentPeriodTextOnly = () => {
    const { startDate, endDate } = periods[0];
    return <strong data-test-id="usage-period-text">{processDate(startDate, endDate)}</strong>;
  };

  return (
    <div className="usage__period-selector">
      <label>API requests usage period</label>
      {isTeamOrEnterpriseCustomer ? renderSelect() : renderCurrentPeriodTextOnly()}
    </div>
  );
};
