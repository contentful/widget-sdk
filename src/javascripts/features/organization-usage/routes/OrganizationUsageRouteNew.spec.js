import React from 'react';
import { shallow } from 'enzyme';
import 'jest-enzyme';
import { Spinner, Select } from '@contentful/forma-36-react-components';
import {
  OrganizationUsageRouteNew,
  WorkbenchContent,
  WorkbenchActions,
} from './OrganizationUsageRouteNew';
import { PeriodSelector } from '../components/PeriodSelector';
import { NoSpacesPlaceholder } from '../components/NoSpacesPlaceholder';
import { OrganizationUsagePage } from '../components/OrganizationUsagePage';
import ReloadNotification from 'app/common/ReloadNotification';
import * as OrganizationRolesMocked from 'services/OrganizationRoles';
import * as TokenStoreMocked from 'services/TokenStore';
import * as OrganizationMembershipRepositoryMocked from 'access_control/OrganizationMembershipRepository';
import {
  isEnterprisePlan,
  getPlansWithSpaces,
  isSelfServicePlan,
} from 'account/pricing/PricingDataProvider';
import createResourceService from 'services/ResourceService';
import { getPeriods, getOrgUsage, getApiUsage } from '../services/UsageService';
import LoadingState from 'app/common/LoadingState';

jest.mock('services/intercom', () => ({}));
jest.mock('utils/ResourceUtils', () => ({}));
jest.mock('services/OrganizationRoles', () => ({
  isOwnerOrAdmin: jest.fn().mockReturnValue(true),
}));

jest.mock('services/ResourceService', () => {
  const service = {
    get: jest.fn((resource) => {
      switch (resource) {
        case 'api_request':
          return { limits: { included: 1000000 } };
        case 'asset_bandwidth':
          return {
            usage: 200,
            unitOfMeasure: 'MB',
            limits: { included: 2000 },
          };
      }
    }),
    getAll: jest.fn(),
  };

  return () => service;
});

jest.mock('account/pricing/PricingDataProvider', () => {
  const getPlansWithSpaces = jest.fn(() => ({
    items: [
      { name: 'Test plan', space: { sys: { id: 'space1' } } },
      { name: 'Proof of concept (space trial)', space: { sys: { id: 'space2' } } },
    ],
  }));
  return {
    isEnterprisePlan: jest.fn(),
    isSelfServicePlan: jest.fn(),
    getBasePlan: jest.fn(),
    getPlansWithSpaces,
  };
});

jest.mock('access_control/OrganizationMembershipRepository', () => ({
  getAllSpaces: jest.fn(() => [
    { name: 'Test1', sys: { id: 'test1' } },
    { name: 'Test2', sys: { id: 'test2' } },
  ]),
}));

jest.mock('services/TokenStore', () => ({
  getOrganization: jest.fn(() => ({})),
}));

jest.mock('data/EndpointFactory', () => {
  const moment = require('moment');
  const _ = require('lodash');
  const DATE_FORMAT = 'YYYY-MM-DD';
  const startDate = moment('2019-01-04').subtract(12, 'days');
  const endpoint = jest.fn().mockImplementation(({ method, path }) => {
    if (method === 'GET' && _.isEqual(path, ['usage_periods'])) {
      return {
        items: [
          {
            startDate: startDate.format(DATE_FORMAT),
            endDate: null,
            sys: { type: 'UsagePeriod', id: '0' },
          },
          {
            startDate: moment(startDate)
              .subtract(1, 'day')
              .subtract(1, 'month')
              .format(DATE_FORMAT),
            endDate: moment(startDate).subtract(1, 'day').format(DATE_FORMAT),
            sys: { type: 'UsagePeriod', id: '1' },
          },
        ],
      };
    }
  });

  return {
    createOrganizationEndpoint: () => endpoint,
  };
});

jest.mock('app/common/ReloadNotification', () => ({
  trigger: jest.fn(),
}));

jest.mock('../services/UsageService', () => ({
  getPeriods: jest.fn().mockReturnValue({ items: [{}] }),
  getApiUsage: jest.fn(),
  getOrgUsage: jest.fn(),
}));

const shallowRenderComponent = async (props) => {
  const wrapper = shallow(<OrganizationUsageRouteNew {...props} />);
  // Need to wait for internal async logic to finish
  // This means `componentDidMount` is called twice,
  //  which could potentially cause problems if it's not
  //  idempotent.
  await wrapper.instance().componentDidMount();
  return wrapper;
};

describe('OrganizationUsageRouteNew', () => {
  let defaultProps;

  beforeAll(() => {
    defaultProps = {
      orgId: '23423',
    };
    // set fixed date for stable snapshots
    // moment('2017-12-01').unix() = 1512082800
    jest.spyOn(Date, 'now').mockImplementation(() => 1512082800);
  });

  afterAll(() => {
    Date.now.mockRestore();
  });

  it('should render page without errors', async () => {
    const wrapper = await shallowRenderComponent(defaultProps);

    expect(wrapper).toMatchSnapshot();
  });

  describe('user is not owner or admin', () => {
    it('should populate error in the state', async () => {
      OrganizationRolesMocked.isOwnerOrAdmin.mockReturnValueOnce(false);

      const wrapper = await shallowRenderComponent(defaultProps);

      expect(TokenStoreMocked.getOrganization).toHaveBeenCalledWith(defaultProps.orgId);
      expect(OrganizationRolesMocked.isOwnerOrAdmin).toHaveBeenCalledWith({});
      expect(wrapper.state('error')).toBeTruthy();
    });
  });

  describe('fetching org data fails with 404', () => {
    it('should populate error in the state', async () => {
      const error404 = new Error('Test error');
      error404.status = 404;

      OrganizationMembershipRepositoryMocked.getAllSpaces.mockRejectedValueOnce(error404);

      const wrapper = await shallowRenderComponent(defaultProps);

      expect(wrapper.state('error')).toBeTruthy();
    });
  });

  describe('fetching org data fails with 403', () => {
    it('should populate error in the state', async () => {
      const error403 = new Error('Test error');
      error403.status = 403;
      OrganizationMembershipRepositoryMocked.getAllSpaces.mockRejectedValueOnce(error403);

      const wrapper = await shallowRenderComponent(defaultProps);

      expect(wrapper.state('error')).toBeTruthy();
    });
  });

  describe('fetching org data fails with different error code', () => {
    it('should trigger reload notification', async () => {
      const error400 = new Error('Test error');
      error400.status = 400;

      OrganizationMembershipRepositoryMocked.getAllSpaces.mockRejectedValueOnce(error400);

      await shallowRenderComponent(defaultProps);

      expect(ReloadNotification.trigger).toHaveBeenCalled();
    });
  });

  describe('org is on the community tier', () => {
    beforeEach(() => {
      isEnterprisePlan.mockReset().mockReturnValue(false);
      isSelfServicePlan.mockReset().mockReturnValue(false);
    });

    it('should fetch data', async () => {
      await shallowRenderComponent(defaultProps);

      expect(getPlansWithSpaces).toHaveBeenCalled();
      expect(OrganizationMembershipRepositoryMocked.getAllSpaces).toHaveBeenCalled();
      expect(getPeriods).toHaveBeenCalled();
      expect(createResourceService().get).toHaveBeenCalledWith('api_request');

      expect(getOrgUsage).toHaveBeenCalled();
      expect(getApiUsage).toHaveBeenCalled();
    });
  });

  describe('org is on the team or enterprise tier', () => {
    describe('enterprise tier', () => {
      beforeEach(() => {
        isEnterprisePlan.mockReset().mockReturnValue(true);
        isSelfServicePlan.mockReset().mockReturnValue(false);
      });

      it('should fetch data', async () => {
        await shallowRenderComponent(defaultProps);

        expect(getPlansWithSpaces).toHaveBeenCalled();
        expect(OrganizationMembershipRepositoryMocked.getAllSpaces).toHaveBeenCalled();
        expect(createResourceService().get).toHaveBeenCalledWith('api_request');
        expect(getPeriods).toHaveBeenCalled();

        expect(getOrgUsage).toHaveBeenCalled();
        expect(getApiUsage).toHaveBeenCalled();
      });
    });

    describe('team tier', () => {
      beforeEach(() => {
        isEnterprisePlan.mockReset().mockReturnValue(false);
        isSelfServicePlan.mockReset().mockReturnValue(true);
      });

      it('should fetch data', async () => {
        await shallowRenderComponent(defaultProps);

        expect(getPlansWithSpaces).toHaveBeenCalled();
        expect(OrganizationMembershipRepositoryMocked.getAllSpaces).toHaveBeenCalled();
        expect(createResourceService().get).toHaveBeenCalledWith('api_request');
        expect(getPeriods).toHaveBeenCalled();

        expect(getApiUsage).toHaveBeenCalled();
        expect(getOrgUsage).toHaveBeenCalled();
      });
    });
  });
});

describe('WorkbenchActions', () => {
  let defaultProps;

  beforeEach(() => {
    defaultProps = {
      hasSpaces: true,
      periods: [],
      selectedPeriodIndex: 0,
      setPeriodIndex: jest.fn(),
      isAssetBandwidthTab: false,
      isTeamOrEnterpriseCustomer: true,
    };
  });

  describe('isLoading', () => {
    it('should render a spinner', () => {
      const wrapper = shallow(<WorkbenchActions isLoading={true} {...defaultProps} />);

      expect(wrapper.find(Spinner)).toHaveLength(1);

      expect(wrapper.find(PeriodSelector)).toHaveLength(0);
    });

    it('should not render spinner if there are no spaces', () => {
      const wrapper = shallow(<WorkbenchActions isLoading hasSpaces={false} />);

      expect(wrapper.find(Spinner)).toHaveLength(0);

      expect(wrapper.find(PeriodSelector)).toHaveLength(0);
    });
  });

  describe('org is on the enterprise or team tier', () => {
    it('should render the PeriodSelector', () => {
      const wrapper = shallow(<WorkbenchActions {...defaultProps} />);

      expect(wrapper.find(PeriodSelector)).toHaveLength(1);
    });

    it('should render nothing if there are no spaces', () => {
      defaultProps.hasSpaces = false;
      const wrapper = shallow(<WorkbenchActions {...defaultProps} />);

      expect(wrapper.find(PeriodSelector)).toHaveLength(0);
      expect(wrapper.find(Spinner)).toHaveLength(0);
    });
  });

  describe('org is on the community tier', () => {
    it('should render the PeriodSelector without Select', () => {
      defaultProps.isTeamOrEnterpriseCustomer = false;
      const wrapper = shallow(<WorkbenchActions {...defaultProps} />);

      expect(wrapper.find(PeriodSelector)).toHaveLength(1);
      expect(wrapper.find(Select)).toHaveLength(0);
    });

    it('should render nothing if there are no spaces', () => {
      defaultProps.isTeamOrEnterpriseCustomer = false;
      defaultProps.hasSpaces = false;
      const wrapper = shallow(<WorkbenchActions {...defaultProps} />);

      expect(wrapper.find(PeriodSelector)).toHaveLength(0);
      expect(wrapper.find(Spinner)).toHaveLength(0);
    });
  });
});

describe('WorkbenchContent', () => {
  let defaultProps = null;

  beforeEach(() => {
    defaultProps = {
      hasSpaces: true,
      selectedPeriodIndex: 0,
      spaceNames: { space1: 'Space1', space2: 'Space2' },
      isPoC: { space1: false, space2: true },
      periodicUsage: {
        org: { usage: [] },
        apis: { cma: { items: [] } },
      },
      apiRequestIncludedLimit: 1000,
      assetBandwidthUsage: 100,
      assetBandwidthIncludedLimit: 50,
      assetBandwidthUOM: 'GB',
      isLoading: false,
      periods: [],
    };
  });

  describe('isLoading', () => {
    it('should render a LoadingState', () => {
      defaultProps.isLoading = true;
      const wrapper = shallow(<WorkbenchContent {...defaultProps} />);

      expect(wrapper.find(LoadingState)).toHaveLength(1);
    });
  });

  it('should render the OrganizationUsagePage if there are spaces', () => {
    const wrapper = shallow(<WorkbenchContent {...defaultProps} />);

    expect(wrapper.find(OrganizationUsagePage)).toHaveLength(1);

    expect(wrapper.find(NoSpacesPlaceholder)).toHaveLength(0);
  });

  it('should render NoSpacePlaceholder if there are no spaces', () => {
    defaultProps.hasSpaces = false;
    const wrapper = shallow(<WorkbenchContent {...defaultProps} />);

    expect(wrapper.find(NoSpacesPlaceholder)).toHaveLength(1);

    expect(wrapper.find(OrganizationUsagePage)).toHaveLength(0);
  });
});
