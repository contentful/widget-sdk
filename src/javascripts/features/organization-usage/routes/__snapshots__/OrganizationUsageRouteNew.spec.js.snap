// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrganizationUsageRouteNew should render page without errors 1`] = `
.emotion-0 {
  height: 100%;
}

.emotion-0 > div {
  height: 100%;
}

<Fragment>
  <DocumentTitle
    title="Usage"
  />
  <Workbench
    testId="organization.usage"
  >
    <WorkbenchHeader
      actions={
        <WorkbenchActions
          hasSpaces={true}
          isAssetBandwidthTab={false}
          isLoading={false}
          periods={
            Array [
              Object {},
            ]
          }
          setPeriodIndex={[Function]}
        />
      }
      icon={
        <NavigationIcon
          color="green"
          icon="usage"
          size="large"
        />
      }
      testId="cf-ui-workbench-header"
      title="Usage"
    />
    <WorkbenchContent
      className="emotion-0"
      testId="cf-ui-workbench-content"
      type="default"
    >
      <WorkbenchContent
        apiRequestIncludedLimit={1000000}
        error={null}
        hasSpaces={true}
        isLoading={false}
        isPoC={
          Object {
            "space1": false,
            "space2": false,
          }
        }
        onTabSelect={[Function]}
        periods={
          Array [
            Object {},
          ]
        }
        spaceNames={
          Object {
            "test1": "Test1",
            "test2": "Test2",
          }
        }
      />
    </WorkbenchContent>
  </Workbench>
</Fragment>
`;
